source/tutorials/learning-journey/packaging-existing-software.md

```diff
@@ -0,0 +1,531 @@
+
+To add this step to your derivation, use the [`installPhase` attribute](https://nixos.org/manual/nixpkgs/stable/#ssec-install-phase), which contains a list of command strings to execute to perform the installation.
+
+Because the `make` step completes successfully, the `icat` executable is available in the build directory, and you only need to copy it from there to the output directory.
+
+In Nix, this location is stored in the `$out` variable, accessible in the derivation's component scripts; create a `bin` directory within that and copy the `icat` binary there:
+
+```nix
+# icat.nix
+{ lib
+, stdenv
+, fetchFromGitHub
+, imlib2
+, xorg
+}:
+
+stdenv.mkDerivation {
+  name = "icat";
+
+  src = fetchFromGitHub {
+    owner = "atextor";
+    repo = "icat";
+    rev = "v0.5";
+    sha256 = "0wyy2ksxp95vnh71ybj1bbmqd5ggp13x3mk37pzr99ljs9awy8ka";
+  };
+
+  buildInputs = [ imlib2 xorg.libX11.dev ];
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp icat $out/bin
+  '';
+}
+```
+
+### Phases and Hooks
+Nixpkgs `stdenv.mkDerivation` derivations are separated into [phases](https://nixos.org/manual/nixpkgs/stable/#sec-stdenv-phases), each of which is intended to control some aspect of the build process.
+
+You saw earlier how the `stdenv.mkDerivation` expected the project's `Makefile` to have an `install` target, and failed when it didn't.
+
+To fix this, you defined a custom `installPhase`, containing instructions for copying the `icat` binary to the correct output location, in effect installing it.
+
+Up to that point, the `stdenv.mkDerivation` automatically determined the `buildPhase` information for the `icat` package.
+
+During derivation realisation, there are a number of shell functions ("hooks", in `nixpkgs`) which may execute in each derivation phase, which do things like set variables, source files, create directories, and so on.
+
+These are specific to each phase, and run both before and after that phase's execution, controlling the build environment and helping to prevent environment-modifying behavior defined within packages from creating sources of nondeterminism within and between Nix derivations.
+
+It's good practice when packaging software with Nix to include calls to these hooks in the derivation phases you define, even when you don't make direct use of them; this facilitates easy [overriding](https://nixos.org/manual/nixpkgs/stable/#chap-overrides) of specific parts of the derivation later, in addition to the previously-mentioned reproducibility benefits.
+
+You should now adjust your `installPhase` to call the appropriate hooks:
+
+```nix
+# icat.nix
+...
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin
+    cp icat $out/bin
+    runHook postInstall
+  '';
+...
+```
+
```

Comment:
```suggestion
### A successful build
```

Finally building the package is an important step and shouldn't be (accidentally) hidden inside of the "Phases and Hooks" section, where it can be overlooked.